### PR TITLE
Add complexity annotations to std. library

### PR DESCRIPTION
**Pull request by esbenbjerre**
WIP

### CODE REVIEW

File: main/src/library/BigInt.flix [hunk 1]
```diff
@@ -19,26 +19,31 @@ namespace BigInt {
     ///
```
Comment (magnus-madsen) [inline main/src/library/BigInt.flix:19]: Just a quick comment: For arbitrary-sized arithmetic the operations are never constant size nor space. I suggest we simply leave them un-annotated for now.

Comment (esbenbjerre) [reply main/src/library/BigInt.flix:19]: Roger that.

**Comment by magnus-madsen** [comment]
Let me know when I should take a pass through the code.

**Comment by esbenbjerre** [comment]
I'm unsure about the complexity of `ref` and `deref`.
I assume `ref` is O(1) and simply returns a reference to the object, but I'm unsure how deref works, e.g. looking at `MutMap. toImmutable`:
```
pub def toImmutable(m: MutMap[k, v]): Map[k, v] & Impure =
    let MutMap(mm) = m;
    deref mm
```
Does dereferencing in this context equal copying the object?

**Comment by magnus-madsen** [comment]
> 
> 
> I'm unsure about the complexity of `ref` and `deref`.
> I assume `ref` is O(1) and simply returns a reference to the object, but I'm unsure how deref works, e.g. looking at `MutMap. toImmutable`:
> 
> ```
> pub def toImmutable(m: MutMap[k, v]): Map[k, v] & Impure =
>     let MutMap(mm) = m;
>     deref mm
> ```
> 
> Does dereferencing in this context equal copying the object?

No, there is no copying. Both are O(1).

**Comment by esbenbjerre** [comment]
> > I'm unsure about the complexity of `ref` and `deref`.
> > I assume `ref` is O(1) and simply returns a reference to the object, but I'm unsure how deref works, e.g. looking at `MutMap. toImmutable`:
> > ```
> > pub def toImmutable(m: MutMap[k, v]): Map[k, v] & Impure =
> >     let MutMap(mm) = m;
> >     deref mm
> > ```
> > 
> > 
> > Does dereferencing in this context equal copying the object?
> 
> No, there is no copying. Both are O(1).

Roger. Another problem related to MutMap
```
-- Type Error -------------------------------------------------- MutMap.flix

>> Unable to unify the Boolean formulas: 'true' and 'false'.

49 |     @Time(Int32.log2(size(m))) @Space(1)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
         mismatched boolean formulas.

Type One: true
Type Two: false

If the Boolean formula describes purity:

  (1) Did you forget to mark the function as impure?
  (2) Are you trying to pass a pure function where an impure is required?
  (3) Are you trying to pass an impure function where a pure is required?

If the Boolean formula describes nullability:

  (1) Are you trying to pass null where a non-null value is required?
```
This seems to be because `MutMap.size` is impure. Whould you prefer me to cast the calls to `MutMap.size as & Pure` or leave `MutMap` and `MutSet` un-annotated for now?

**Comment by magnus-madsen** [comment]
> 
> 
> > > I'm unsure about the complexity of `ref` and `deref`.
> > > I assume `ref` is O(1) and simply returns a reference to the object, but I'm unsure how deref works, e.g. looking at `MutMap. toImmutable`:
> > > ```
> > > pub def toImmutable(m: MutMap[k, v]): Map[k, v] & Impure =
> > >     let MutMap(mm) = m;
> > >     deref mm
> > > ```
> > > 
> > > 
> > > Does dereferencing in this context equal copying the object?
> > 
> > 
> > No, there is no copying. Both are O(1).
> 
> Roger. Another problem related to MutMap
> 
> ```
> -- Type Error -------------------------------------------------- MutMap.flix
> 
> >> Unable to unify the Boolean formulas: 'true' and 'false'.
> 
> 49 |     @Time(Int32.log2(size(m))) @Space(1)
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^
>          mismatched boolean formulas.
> 
> Type One: true
> Type Two: false
> 
> If the Boolean formula describes purity:
> 
>   (1) Did you forget to mark the function as impure?
>   (2) Are you trying to pass a pure function where an impure is required?
>   (3) Are you trying to pass an impure function where a pure is required?
> 
> If the Boolean formula describes nullability:
> 
>   (1) Are you trying to pass null where a non-null value is required?
> ```
> 
> This seems to be because `MutMap.size` is impure. Whould you prefer me to cast the calls to `MutMap.size as & Pure` or leave `MutMap` and `MutSet` un-annotated for now?

Actually, maybe you should just defer MutMap- because its implementation is likely to change :)

**Comment by magnus-madsen** [comment]
Could you post a list of namespaces that are done and a list that you are planning to do/working on?

**Comment by esbenbjerre** [comment]
> Could you post a list of namespaces that are done and a list that you are planning to do/working on?

Yes.
- [x] Array.flix
- [x] ~~~BigInt.flix~~~
- [x] Bool.flix
- [x] Bounded.flix
- [x] Char.flix
- [x] Float32.flix
- [x] Float64.flix
- [x] Int8.flix
- [x] Int16.flix
- [x] Int32.flix
- [x] Int64.flix
- [x] List.flix
- [x] Map.flix
- [x] MutList.flix
- [x] MutMap.flix
- [ ] MutSet.flix
- [ ] Nel.flix
- [x] Object.flix
- [x] Option.flix
- [x] Prelude.flix
- [x] Random.flix
- [x] RedBlackTree.flix
- [ ] Result.flix
- [x] Set.flix
- [ ] String.flix
- [ ] Validation.flix

**Comment by magnus-madsen** [comment]
Excellent. Let us skip (remove):

- Bounded
- MutList, MutMap, MutSet (we can think about those for later)
- Object
- Random

I know you did some of them already, but I think it is very little code being removed. I just want to have a small core of namespaces to begin with. 

I will think about the mutability..

**Comment by esbenbjerre** [comment]
Sounds reasonable :)
I'll focus on the remaining.

File: main/src/library/Array.flix [hunk 2]
```diff
@@ -270,13 +292,15 @@ namespace Array {
...
     ///
+    @Time(time(f) * length(a)) @Space(space(f) * length(a))
     pub def map(f: a -> b & e, a: Array[a]): Array[b] & Impure =
         let len = length(a);
         init(i -> f(a[i]), len)
 
     ///
     /// Apply `f` to every element in array `a`. Array `a` is mutated.
     ///
+    @Time(time(f) * length(a)) @Space(space(f))
```
Comment (magnus-madsen) [inline main/src/library/Array.flix:303]: I think the space needs to be multiplied by the length

File: main/src/library/Array.flix [hunk 3]
```diff
@@ -335,6 +363,7 @@ namespace Array {
     ///
     /// Reverse the array `a`, mutating it in place.
     ///
+    @Time(length(a)) @Space(1)
     pub def reverse!(a: mut Array[a]): Unit & Impure =
```
Comment (magnus-madsen) [inline main/src/library/Array.flix:367]: I like how the annotations illustrate the difference between `reverse` and `reverse!`. Beautiful :)

File: main/src/library/Array.flix (hunk 2 above)
Comment (magnus-madsen) [reply main/src/library/Array.flix:303]: My reasoning is that if every element is replaced by a potentially much larger element from f.

Comment (magnus-madsen) [reply main/src/library/Array.flix:303]: If you have an array of empty strings and replace every element by a very large string then you know have an array of large strings.

File: main/src/library/List.flix [hunk 4]
```diff
@@ -402,6 +406,7 @@ namespace List {
     ///
     /// That is, returns `y1 :: x1 ... xn :: y2 :: ... yn-1 :: x1 :: ... :: xn :: yn :: Nil`.
     ///
+    @Time(length(ys)) @Space(length(ys))
     pub def intercalate(xs: List[a], ys: List[List[a]]): List[a] = match ys {
```
Comment (magnus-madsen) [inline main/src/library/List.flix:410]: I think this is wrong. Imagine if `xs` is very large.

**Comment by magnus-madsen** [comment]
If we have functions like folds, reduces, etc:

```
    @Time(time(f) * length(xs)) @Space(space(f))
    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: List[a]): b & e = match xs {
        case Nil => s
        case x :: rs => foldLeft(f, f(s, x), rs)
    }
```

then I think the space usage has to be higher. Let us try an example:

```
def f(_x: Int, _y: Int): Int = 1 // Space is O(1)
def foo(): List[Int] = foldLeft(f, 0, veryLargeList)
```

If I understand it correctly, I would say that the space usage of foldLeft here is not O(1) as suggested by the annotation on foldLeft. Rather it should be O(space(f) * length(l)). Because, then, here it would be O(length(l)) as expected.

On the other hand, this signature:

```
   @Time(time(f) * length(xs)) @Space(space(f))
    pub def exists(f: a -> Bool, xs: List[a]): Bool = match xs {
```

seems perfectly fine - because whenever we invoke f it can only use space(f) space. And the result is always a Boolean (of size O(1)) so there is no way to require more space than needed by `f`.

Does it make sense?

**Comment by magnus-madsen** [comment]
That said, I actually worry that my analysis is also too optimistic. What if we have:

```
@Time(n) @Space(n)
def double(l: List[Int]): List[Int] = ... // simply doubles the length of the list

def f(): List[Int] = foldLeft((x, xs) -> double(x :: xs), Nil, largeList)
```

This doubles the list for every element in the original list. Hence the total length should be n^2 (agree?) But our annotations arrives at significantly faster time and space. The problem is really that when we fold we can construct increasingly larger inputs.

**Comment by esbenbjerre** [comment]
That makes sense. I will take another look at these.
I'll have to think about fold, but I do see the issue.

**Comment by magnus-madsen** [comment]
> 
> 
> That makes sense. I will take another look at these.
> I'll have to think about fold, but I do see the issue.

I wrote to a colleague to ask for help :)

Either we can figure it out quickly or it may be that we leave the folds for later.

**Comment by magnus-madsen** [comment]
Note that for e.g. `map` the analysis is easier, you just apply the function to every element so clearly there is not this possibility of feeding the result back into itself... If that makes sense.. That is the scenario you should look for.

**Comment by esbenbjerre** [comment]
Roger.
W.r.t the functions in `String.flix` which uses Java interop, should I skip these or look at take the Java implementation into account in order to analyze them?

**Comment by magnus-madsen** [comment]
> 
> 
> Roger.
> W.r.t the functions in `String.flix` which uses Java interop, should I skip these or take the Java implementation into account in order to analyze them?

You should think of them as immutable array of characters.

So, for example, appending takes time in the size of both strings.

Looking up an index is O(1).

**Comment by magnus-madsen** [comment]
I talked to a colleague. Based on our discussion, I suggest that we omit (remove) all time and space complexities from: folds, reduces, scans. Moreover, let us skip: sequence and traverse + unfolds. We will revisit these later once we have a better grasp of the situation. Meanwhile, I suggest we wrap up this PR and merge it in, and then we can move to other tasks afterwards.

**Comment by esbenbjerre** [comment]
All right.
I've added annotations to String.flix until line 895.
Would you prefer me to add the remaining or do you have a task with higher priority?

**Comment by magnus-madsen** [comment]
I think finish up the namespaces we decided on, including string. But leave out the folds etc.

**Comment by esbenbjerre** [comment]
> I think finish up the namespaces we decided on, including string. But leave out the folds etc.

I think that's it. I've also removed the annotations from folds etc.

File: main/src/library/Set.flix [hunk 5]
```diff
@@ -30,6 +30,7 @@ namespace Set {
     ///
     /// Returns the size of `xs`.
     ///
+    @Time(size(xs)) @Space(Int32.log2(size(xs)))
```
Comment (magnus-madsen) [inline main/src/library/Set.flix:33]: I mean, is this not practically O(1) space for all purposes?

**Review by magnus-madsen — COMMENTED** [review]
I added a few comments, but essentially I think this is ready to be merged in.

File: main/src/library/Set.flix (hunk 5 above)
Comment (esbenbjerre) [reply main/src/library/Set.flix]: That is a fair point.

File: main/src/library/List.flix [hunk 6]
```diff
@@ -877,13 +906,15 @@ namespace List {
     ///
     /// Returns the list of characters `xs` as a string.
     ///
+    @Time(length(xs)) @Space(1)
     pub def toString(xs: List[Char]): Str =
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/List.flix:910]: You're absolutely right.
As per "Str" I vaguely recall you saying something about it being an old type name. But I do not see that I have been assigned a specific task about it. A quick IntelliJ search reveals multiple places where it's still being used. Maybe we should create a separate issue for this?

**Comment by magnus-madsen** [comment]
Is this ready?

**Comment by esbenbjerre** [comment]
It is.

**Pull request merged**